packages/grid/x-data-grid/src/internals/hooks/features/pagination/useGridPage.ts

```diff
@@ -160,9 +168,25 @@ export const useGridPage = (
   /**
    * EFFECTS
    */
+  React.useEffect(() => {
+    if (process.env.NODE_ENV !== 'production') {
+      if (props.paginationMode === 'server' && props.rowCount == null) {
+        noRowCountInServerMode();
+      }
+    }
+  }, [props.rowCount, props.paginationMode]);
+
   React.useEffect(() => {
     apiRef.current.setState((state) => {
```

Comment:
I suggest to: 

- Remove this logic
- Add a doc note on the pagination server saying that if the `rowCount` switch to undefined while loading, users must to the trick you implemented here
- Implement the same trick but in our server pagination demo (and change the fake server behavior to return "undefined" while loading
- Change the warning to link to the doc note

Reply:
🤯 That's a mind-blowing suggestion: 
creating a horrible demo to let users know how to do 😄

Reply:
"Horrible", it's the Apollo Client behavior :laughing: 
But alternatively, we can add a new "Apollo Client" section. It would make sense to start giving examples with various famous data management libraries.